clients/go/events/idempotent_recorder.go

```diff
@@ -0,0 +1,96 @@
+package events
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+
+	eventsErr "github.com/lyft/flyteidl/clients/go/events/errors"
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/event"
+
+	"github.com/lyft/flytestdlib/logger"
+	"github.com/lyft/flytestdlib/promutils"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+type EmittedEventInfo struct {
+	lastUpdatedAt time.Time
+	phase         string
+}
+
+type idempotentWorkFlowEventRecorder struct {
+	internalRecorder    *eventRecorder
+	maxUpdateLagSeconds int64
+	cache               *lru.Cache
+}
+
+func (i *idempotentWorkFlowEventRecorder) idempotentRecord(ctx context.Context, id string, e proto.Message, phase string) error {
+	var existing *EmittedEventInfo
+	o, found := i.cache.Get(id)
+	if found {
+		existing = o.(*EmittedEventInfo)
+	}
+
+	emitTime := time.Now()
+	if !found || existing.phase != phase || emitTime.Sub(existing.lastUpdatedAt) > time.Duration(i.maxUpdateLagSeconds)*time.Second {
```

Comment:
Should we attempt to make it a bit more involved here? e.g. for Array tasks, the task might move to running before we have logs for its subtasks, when they start being constructed, with this change, we will only see the update once every 5 mins...
The alternative might be to also cache the phase version... because that we do change every time we change something...

Should we make the check based on the hashed value of the PhaseInfo we are trying to send or something of that sort?